src/lib.rs

```diff
@@ -32,9 +32,42 @@
 mod windows;
 
 use std::env;
+use std::ffi::OsString;
 use std::io;
 use std::path::{Path, PathBuf};
 
+/// Permits parameterising the home functions via the _from variants - used for
+/// in-process unit testing by rustup.
+pub trait Env {
+    fn home_dir(&self) -> Option<PathBuf>;
+    fn current_dir(&self) -> io::Result<PathBuf>;
+    fn var_os(&self, key: &str) -> Option<OsString>;
+}
```

Comment:
What if we use default implementation? That's way we don't need to public OsEnv.

Reply:
We use OsEnv here: https://github.com/rust-lang/rustup/blob/master/src/currentprocess/homethunk.rs#L37